### ISSUE TITLE
Merging config with multiple targets

### ISSUE THREAD

**Issue by herecydev** [survivejs/webpack-merge#72]
I'm after using [multiple targets](https://webpack.js.org/concepts/targets/) to have a server side config and a client side config packaged. Where one targets node and the other uses the default web target.

webpack.base.js
```javascript
let clientConfig = {
  entry: "./src/website/js/App.tsx"
}

let serverConfig = {
  entry: "./src/server/server.js",
  output: {
    path: path.resolve(__dirname, "dist"),
    filename: "[name].js",
  },
  target: "node"
};
}
```

webpack.config.js
```javascript
let clientConfig = {
  devtool: "source-map"
};

module.exports = webpackMerge(config, [clientConfig]);
```

Before the merge the json looks healthy:

```
[ { entry: './src/website/js/App.tsx' },
  { entry: './src/server/server.js',
    output: { path: 'C:\\Dev\\dist', filename: '[name].js' },
    target: 'node' } ]
```

But the output from the merge looks like it's become a single object and hence now fails with the typical errors `'output.filename' is required ...`

```
{ '0':
   { entry: './src/website/js/App.tsx',
     devtool: 'source-map',
     module: { rules: [Object] },
     output: { filename: '[name].js' },
     plugins: [ [Object], [Object] ] },
  '1':
   { entry: './src/server/server.js',
     output: { path: 'C:\\Dev\\dist', filename: '[name].js' },
     target: 'node' } }
```

**Comment by bebraw** [comment]
The problem is the mixed types at `webpackMerge(config, [clientConfig]);`. We probably want to disallow mixing objects with arrays (raise `Error`?) for the root.

The right way to merge in this case would be to do either `webpackMerge(config, clientConfig);` or `webpackMerge([config, clientConfig]);`.

**Comment by herecydev** [comment]
Neither of those work though, you're still left with an object with the keys 0, 1, etc. Which I assume webpack doesn't understand (when it looks for the filename, entrypoints, etc).

**Comment by bebraw** [comment]
Ah... I got the problem now. Basically you want to merge `clientConfig` to each array config. The tool was designed under assumption that you work with a single configuration at a time.

Based on that you could end up with something like this:

```javascript
const serverConfigs = [{...}, {...}]; // Array of configs
const clientConfig = {...}; // Single client config

// Merge each client config to server configs
const config = serverConfigs.map(
  serverConfig => merge(serverConfig, clientConfig)
);
```

Would this work for you? I could add an example to the readme.

**Comment by herecydev** [comment]
Here's probably the most succinct gist I can think of: https://gist.github.com/herecydev/af366b06e93a1aeacd53f80ebad01633

**Comment by bebraw** [comment]
Ok. That might call for a little different composition. It can helpful to try something along this:

```javascript
module.exports = [
  merge(commonConfig, serverConfig),
  merge(commonConfig, clientConfig)
];
```

This is just a rough example, but if you extract the array concern to your webpack configuration, you'll get something that's easy to reason with. I hope this gets closer to what you need. 👍

**Comment by herecydev** [comment]
Yer I think that works, problem is, it's not the most DRY solution right? All of the webpack configuration files have to know to include all of the targets, as opposed to each configuration file changing only what's necessary.

**Comment by bebraw** [comment]
Can you expand the gist with more actual config and references between the files? Make sure to mark what's out of my bounds. I don't mind giving it another go.

**Comment by herecydev** [comment]
Updated the gist, 1st sample is the base config, 2nd sample is the ideal approach (I'm only altering the targets I care about), 3rd is a less ideal scenario for a dev environment, 4th is a production environment. In both scenario 3 and 4 I have to explicitly redefine my exports.

**Comment by bebraw** [comment]
Thanks. I think I see a way that could be a good compromise:

1. Change **base.js** to export like `module.exports = { clientConfig, serverConfig }`.
2. **dev.js** - `module.exports = [merge(base.clientConfig, clientConfig), base.serverConfig];`
3. **prod.js** - `module.exports = [base.clientConfig, base.serverConfig];`

It's a bit more code than in the ideal solution. The key insight is that **base.js** doesn't have to care about composition at all. It's enough if it exposes the configuration fragments you compose later. In production you can skip merging entirely as earlier. During development you tweak the client configuration. If you needed to tweak the server too, it would be possible there.

I thought about expanding **webpack-merge** to support what you need but that would lead to a special case to document (also something for users to remember). That said, it could be sensible to write a multi-compiler friendly variant which performs merge based on array position like in your example.

**Comment by herecydev** [comment]
Yes, in my code it is structured like that. I'm sure you can see the problem when scaled up. Imagine 10 configurations with 10 targets. Each configuration file should have an easy way to amend a specific target without having to recompose the targets. Refactoring becomes more complicated.. I think you get the picture. I'm not sure how much multi-targets are used, but in this case I had a React app and a node backend for which I had to start introducing more targets for.

**Comment by bebraw** [comment]
@herecydev Yup. That actually brings another problem. Merging arrays has a positional dependency inside. If the position changes somehow, you break configuration. I think we could figure out a nice solution for this use case, though.

From webpack point of view the order of the entries does not matter. It will compile them anyway. It is enough that we emit an array. Based on this I would go with an API that skips the position problem and gives you this output. Example:

```javascript
const base = {
  client: {... client config ...},
  server: {... server config ...},
  ... add more keys as needed ...
};

const dev = {
  client: {... specialize client config now ...},
  ... specialize more as needed ...
};

// The trick is here. merge.multiple takes the object and turns them into
// [{... merged client config ...}, {... merged server config ...}]
module.exports = merge.multiple(base, dev);
```

`merge.multiple` would essentially work as a tiny adapter. It would not have to do that much.

Would this kind of API work for you? Any better ideas?

**Comment by herecydev** [comment]
No that's perfect, I would imagine any configs that don't exist in the base classes will just be added, otherwise they go through the normal merging process. Wrapping a single configuration in an object (essentially a dictionary) isn't ideal but I see no other option.

**Comment by bebraw** [comment]
The object is to deal with the problem of order, yeah. There's a little aliasing problem now that I think of it (configurations can have different amounts of configurations) but that's solvable.

Do you want to PR something? I can give it a go at some point too. The specification itself feels clear.

**Comment by herecydev** [comment]
Added [a pr] for this, I think it achieves what we're after if that lines up with your intentions?

**Comment by bebraw** [comment]
Nice start! I added a comment as there's are two points to test and specify.

**Comment by ggirard07** [comment]
Can this be considered closed as #74 got merged in a while back?

**Comment by bebraw** [comment]
@ggirard07 Yeah, let's close. Good call.
